#258 refactor flux implementation

Ma27 opened the pull request:
### Overview

fixes certain mistakes in the frontend.
The following descisions have been reverted;
- `flux` as only tool to implement the flux pattern
- abstract component API (caused re-renderings of the menu bar and the whole container, new approach re-renders the content and keeps the menu)

**Done:**
- [x] use `edge` image rather than precise and NodeJS 6 in the build environment
- [x] major dependency bump (e.g. ReactJS 15, react-router 2.0, JSDom 8)
- [x] simplify service structure in `util/http`
- [x] lightweight form API to reduce boilerplate
- [x] structure refactoring to use containers rather than abstract classes to generate boiler-plate JSX (fixes issue with loclae initialization)
- [x] use flux-container to reduce boilerplate in the business logic
- [x] minor refactorings
### Related tickets

resolves #258

Ma27 committed:
Improvement #258 upgrade and improve setup/build/tests of the frontend

- bump dependencies (reactjs to v15, react-router v2, jsdom v8)
- introduce enzyme for shallow-rendering based tests
- use edge image in CI builds to build the frontend using NodeJS v6

Ma27 committed:
Improvement #258 simplify complex service structure

- removed complex facade/cookie API
- rely on localStorage for API keys
- direct usage of `cookies-js` in the internal locale API

Ma27 committed:
Improvement #258 implemented a more lightweight form API to reduce all the boilerplate code

- simple form helper and form field API written in ReactJS code to reduce boiler-plate code
- backend API to return multi-language responses in order to be able to translate server messages without scheduling request multiple times

Ma27 committed:
Improvement #258 entire structure refactoring

- create a container component for the app rather than abstract classes which caused re-renderings of the whole structure
- refactor app component structure
- minor i18n API refactorings

Ma27 committed:
Improvement #258 use the flux-container

Ma27 commented:
before merging, the commits must be squashed in chunks (too much right now, but squash isn't a good idea)

Ma27 commented on .gitignore:
```diff
@@ -78,3 +78,6 @@ npm-debug.log
 /deploy/config/deploy.yaml
+
+# LocalStorage Emulation
+./scratch/*
```
I admit it, I've copied it from the docs. The dir must be renamed of course.

Ma27 commented on src/AppBundle/Controller/BaseController.php:
```diff
@@ -62,4 +88,24 @@ protected function getCurrentUser()
+     *
+     * @return string[]
+     */
+    protected function getLocaleKeys()
```
how about _getLocaleShortNames_?

benbieler replied:
Yes, could be usefull...

Ma27 commented on src/AppBundle/View/I18nResponseBuilder.php:
```diff
@@ -0,0 +1,169 @@
+use Symfony\Component\Validator\ConstraintViolationListInterface;
+
+/**
+ * Concrete implementation of a builder which transforms constraint violations into an i18n response.
```
it doesn't build the response, but the format for it

Ma27 commented on src/AppBundle/View/I18nResponseBuilder.php:
```diff
@@ -0,0 +1,169 @@
+ *
+ * @author Maximilian Bosch <maximilian.bosch.27@gmail.com>
+ */
+class I18nResponseBuilder implements I18nResponseBuilderInterface
```
rename to _I18nResponseFormatBuilder_ (see comment above)?

Ma27 commented on src/AppBundle/View/I18nResponseBuilder.php:
```diff
@@ -0,0 +1,169 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function buildTranslatableResponse(
```
rename to _formatTranslatableViolationList_?

Ma27 commented:
@benbieler have to finish my review, but would you approve?

Ma27 commented on package.json:
```diff
@@ -35,30 +36,31 @@
-    "react-interpolate-component": "^0.8",
+    "react": "^15.0.2",
+    "react-bootstrap": "^0.29",
+    "react-dom": "^15.0",
```
should be increased to `^15.0.1` as 15.0 contained a regression bug

Ma27 commented on src/AppBundle/View/I18nResponseBuilder.php:
```diff
@@ -0,0 +1,169 @@
+            );
+        }
+
+        return array_reduce(
```
I'm still not sure whether `array_walk` is better or an imperative foreach approach

Ma27 commented on src/AppBundle/View/I18nResponseBuilder.php:
```diff
@@ -0,0 +1,169 @@
+                    //   ]
+                    // ]
+                    case $useAllLanguages:
+                        foreach ($targetLocales as $locale) {
```
another point where we should evaluate whether the imperative approach or a `array_` function is the better solution

Ma27 commented on src/Frontend/config/counterpart.js:
```diff
@@ -10,15 +10,11 @@
-import en from '../config/languages/en';
+import Locale from '../util/http/LocaleService';
 
+Locale.setLocale(null);
```
requires a comment as setting the locale to `null` may seem a little bit odd during debug

Ma27 commented on src/Frontend/store/handler/filterItemsByVisibility.js:
```diff
@@ -0,0 +1,28 @@
+ */
+export default (items, authData) => {
+  return items.filter(item => {
+    return !('ROLE_ADMIN' === item.role && !authData.is_admin
```
the first condition should be in a new line, too.

Ma27 commented:
except the failing style validation and the comments, it can be chunked properly and finally merged

Ma27 commented:
everythign seems fine now.
Some commits need squash, certain things should be splitted and in the end we have to reorder stuff a bit

Ma27 commented:
ready for now. I've squashed most of the commits and moved the tests and review fixes to the correct commits to avoid those `fixes review comments` commits and keep the history clean.

Ma27 commented on src/Frontend/components/app/layout/Application.js:
```diff
@@ -0,0 +1,33 @@
+ * @returns {React.Element} The markup of the app.
+ */
+const Application = props => {
+  return (
```
**NOTE:** I didn't implement a refresh menu feature which re-renders the menu structure if something changes. Instead it should listen to the login/logout lifecycle to be implemented in #35 and decide based on the given data of the login/logout cycle whether to re-render. This saves unncessary resources that would be needed to re-evaluate the menu items at every hash-change
